cdc/log.cc

```diff
@@ -479,6 +479,134 @@ api::timestamp_type find_timestamp(const schema& s, const mutation& m) {
     throw std::runtime_error("cdc: could not find timestamp of mutation");
 }
 
+// iterators for collection merge
+template<typename T>
+class collection_iterator : public std::iterator<std::input_iterator_tag, const T> {
+    bytes_view _v, _next;
+    size_t _rem = 0;
+    T _current;
+public:
+    collection_iterator(bytes_view_opt v = {}) 
+        : _v(v.value_or(bytes_view{}))
+        , _rem(_v.empty() ? 0 : read_collection_size(_v, cql_serialization_format::internal()))
+    {
+        if (_rem != 0) {
+            parse();
+        }
+    }
+    collection_iterator(const collection_iterator&) = default;
+    const T& operator*() const {
+        return _current;
+    }
+    const T* operator->() const {
+        return &_current;
+    }
+    collection_iterator& operator++() {
+        next();
+        if (_rem != 0) {
+            parse();
+        } else {
+            _current = {};
+        }
+        return *this;
+    }
+    collection_iterator operator++(int) {
+        auto v = *this;
+        ++(*this);
+        return v;
+    }
+    bool operator==(const collection_iterator& x) const {
+        return _v == x._v;
+    }
+    bool operator!=(const collection_iterator& x) const {
+        return !(*this == x);
+    }
+private:
+    void next() {
+        --_rem;
+        _v = _next;
+    }
+    void parse();
+};
+
+template<>
+void collection_iterator<std::pair<bytes_view, bytes_view>>::parse() {
+    assert(_rem > 0);
+    _next = _v;
+    auto k = read_collection_value(_next, cql_serialization_format::internal());
+    auto v = read_collection_value(_next, cql_serialization_format::internal());
+    _current = std::make_pair(k, v);
+}
+
+template<>
+void collection_iterator<bytes_view>::parse() {
+    assert(_rem > 0);
+    _next = _v;
+    auto k = read_collection_value(_next, cql_serialization_format::internal());
+    _current = k;
+}
+
+template<typename Container, typename T>
```

Comment:
Here: not really, but after some ironing and verification maybe these should be promoted to global scope and generalized usage, and then it will be needed in some way or another. Also hurts noone.

Reply:
+1 to generalizing. They seem useful for collections.